### PR TITLE
Fixed a crash in Activities.Attack.Tick

### LINKED ISSUE #593: Exception during GI/Prism Tank attack.
[exception-2019-12-07T174244Z.log](https://github.com/OpenRA/ra2/files/3935379/exception-2019-12-07T174244Z.log)
[debug.log](https://github.com/OpenRA/ra2/files/3935380/debug.log)

### PR DESCRIPTION
**Pull request by Mailaender**
Closes https://github.com/OpenRA/ra2/issues/593.

### CODE REVIEW

**Comment by pchote** [comment]
The `Attack` activity shouldn't be running at all if there are no active attack traits. The activity needs to be canceling itself instead.

**Comment by Mailaender** [comment]
Updated to cancel the activity as well.

File: OpenRA.Mods.Common/Activities/Attack.cs [hunk 1]
```diff
@@ -95,8 +99,11 @@ public override bool Tick(Actor self)
 			if (!targetIsHiddenActor && target.Type == TargetType.Actor)
 			{
 				lastVisibleTarget = Target.FromTargetPositions(target);
-				lastVisibleMaximumRange = attackTraits.Where(x => !x.IsTraitDisabled)
-					.Min(x => x.GetMaximumRangeVersusTarget(target));
+				var activeAttackTraits = attackTraits.Where(x => !x.IsTraitDisabled);
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Attack.cs]: We shouldn't be reaching this branch at all, and adding the guard here is going to cause trouble for #17227.
Bail out in the check at the start.

File: OpenRA.Mods.Common/Activities/Attack.cs [hunk 2]
```diff
@@ -64,8 +64,12 @@ public Attack(Actor self, Target target, bool allowMovement, bool forceAttack, C
 			    || target.Type == TargetType.FrozenActor || target.Type == TargetType.Terrain)
 			{
 				lastVisibleTarget = Target.FromPos(target.CenterPosition);
-				lastVisibleMaximumRange = attackTraits.Where(x => !x.IsTraitDisabled)
-					.Min(x => x.GetMaximumRangeVersusTarget(target));
+				var activeAttackTraits = attackTraits.Where(x => !x.IsTraitDisabled);
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Attack.cs]: Likewise here

File: OpenRA.Mods.Common/Activities/Attack.cs [hunk 3]
```diff
@@ -64,8 +64,12 @@ public Attack(Actor self, Target target, bool allowMovement, bool forceAttack, C
...
+				if (!activeAttackTraits.Any())
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Attack.cs]: Something would have to be quite bogus for this to be false here in the ctor. Crashing is the correct behaviour here.

**Comment by Mailaender** [comment]
Updated with early returns and found another duplicate query.

File: OpenRA.Mods.Common/Activities/Attack.cs (hunk 3 above)
Comment (Mailaender) [reply OpenRA.Mods.Common/Activities/Attack.cs]: Removed the check in the constructor.

File: OpenRA.Mods.Common/Activities/Attack.cs [hunk 4]
```diff
@@ -92,11 +92,17 @@ public override bool Tick(Actor self)
 
 			bool targetIsHiddenActor;
 			target = RecalculateTarget(self, out targetIsHiddenActor);
+			var activeAttackTraits = attackTraits.Where(x => !x.IsTraitDisabled);
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Activities/Attack.cs]: I wonder if we want to `ToArray` or `ToList` this to save enumerations?

Comment (pchote) [reply OpenRA.Mods.Common/Activities/Attack.cs]: I suspect that would actually be worse, but you'd need to profile to know for sure.

Comment (pchote) [reply OpenRA.Mods.Common/Activities/Attack.cs]: We can avoid allocations here by following the pattern we use in other places: change the `attackTraits` definition to `readonly IEnumerable<AttackFrontal> attackTraits;` and initialize it as `attackTraits = self.TraitsImplementing<AttackFrontal>().ToArray().Where(Exts.IsTraitEnabled);`.

Comment (Mailaender) [reply OpenRA.Mods.Common/Activities/Attack.cs]: Good idea. Thanks!

File: OpenRA.Mods.Common/Activities/Attack.cs [hunk 5]
```diff
@@ -92,11 +92,17 @@ public override bool Tick(Actor self)
 
 			bool targetIsHiddenActor;
 			target = RecalculateTarget(self, out targetIsHiddenActor);
+
+			if (!attackTraits.Any())
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Activities/Attack.cs:93]: Does it make sense to perform this check before recalculating the target, since we cancel anyway?

Comment (Mailaender) [reply OpenRA.Mods.Common/Activities/Attack.cs:93]: I don't think so. Honestly I don't remember. It has been months since I submitted this code. Moved it.

**Review by abcdefg30 — APPROVED** [review]
Lgtm.

**Review by deleted-user-1 — APPROVED** [review]
LGTM.

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ff98147ae741223310d2751db1e9abd845e55be4...3be4e1a07ea6283dc6b33da99c1041d5869611ee)

**Pull request merged**